### ISSUE TITLE
Named required parameter called "required" leads to compiling error

### ISSUE THREAD

**Issue by lohanbodevan** [dart-lang/sdk#38588]
_Probably duplicated but i couldn't find any other issue related_.

If i create a class whose parameters are named-required and one of these parameters is called `required` a compile message is returned:
```
Error: This can't be used as metadata; metadata should be a reference to a compile-time constant variable, or a call to a constant constructor.
```

**How to reproduce?**
Class
```dart
import 'package:meta/meta.dart';

class MyInput {
  final String label;
  final bool required;

  MyInput({@required this.label, @required this.required});
}

void main() {
  MyInput a = MyInput(label: 'hello', required: true);
  print(a.label);
}
```

pubspec.yaml
```yaml
name: required_required
description: Example of named required parameter called required
version: 0.0.1
author: John Doe <johndoe@example.com>
homepage: https://pub.dev/packages/meta

environment:
  sdk: ">=1.8.0 <3.0.0"

dependencies:
  meta: ^1.1.7
```

But if i rename the parameter called `required` to `isRequired`, for example, then works fine.
Perhaps was thought to be like this, but i found hard and took sometime to understand what the error means.

**Comment by bwilkerson** [comment]
I agree that the message needs to be improved. Analyzer should notice that the reference would be valid if the name the user intended to reference were not being shadowed.

However, in this case I believe that the code is correct and the diagnostic should not have been reported.

The specification states in section 9.2 on formal parameters to functions:

> This means that in a generic function declaration, the return type and parameter type annotations can use the formal type parameters, but **the formal parameters are not in scope in the signature**.

(emphasis added) I believe that the formal parameter list is part of the signature, so I interpret that to mean that analyzer should not have resolved the name in the annotation to the formal parameter but instead should have resolved it to the top-level variable from the meta package.

@eernstg @leafpetersen @lrhn @munificent for confirmation.

**Comment by lohanbodevan** [comment]
Thank for the explanation @bwilkerson

**Comment by eernstg** [comment]
Agreed, based on `The current scope for the function’s signature is the scope that encloses the formal parameter scope` which occurs just before the commentary where it says 'the formal parameters are not in scope...'.

**Comment by lrhn** [comment]
For the record, the code is not *correct*, even though the diagnostic might be the wrong one.
```dart
class MyInput {
  final String label;
  final bool required;

  MyInput({@required this.label, @required this.required});
}
```
Here the `@required` would still refer to the `final bool required` declaration, which is not const.

**Comment by eernstg** [comment]
This is one of those rare cases where it does actually matter whether there's a 'static' as well as an 'instance' scope in a class. The specification has [both](https://github.com/dart-lang/language/blob/6df1612f667f7fdd28dd3b66a89e06b75ea98d2a/specification/dartLangSpec.tex#L2129), and this means that `required` in `@required` would actually resolve to the top level declaration (which is imported in the original example, but it still ends up in the library scope).

However, the analyzer and the common front end do not implement this distinction. Moreover, it seems unhelpful to have two distinct scopes in a class body, and the example in this issue shows how confusing and subtle the situation could be if we insist that the `@required` in a constructor argument should be allowed, but it should be an error in a parameter of an instance method (because the former is in the static scope and the latter is in the instance scope, so only the latter can see the instance variable named `required`).

Hence, I proposed in dart-lang/language#550 that we should collapse the static and the instance scope to a single scope (designated as the 'class' scope, following the proposal from @munificent).

I think we should just get it done now, so I created dart-lang/language#1400 where the language specification is adjusted to make this happen. With that in place, we can close this issue with "working as intended".
